activation/activation.go

```diff
@@ -135,133 +139,69 @@ func (b *Builder) Start() {
 
 // Stop stops the atx builder.
 func (b *Builder) Stop() {
-	b.finished <- 0
 	close(b.stop)
 }
 
-func (b Builder) SignAtx(atx *types.ActivationTx) (*types.ActivationTx, error) {
-	bts, err := types.InterfaceToBytes(atx.InnerActivationTx)
-	if err != nil {
-		return nil, err
-	}
-	atx.Sig = b.Sign(bts)
-	return atx, nil
+func (b Builder) SignAtx(atx *types.ActivationTx) error {
+	return SignAtx(b, atx)
 }
 
+type StopRequestedError struct{}
+
+func (s StopRequestedError) Error() string { return "stop requested" }
+
 // loop is the main loop that tries to create an atx per tick received from the global clock
 func (b *Builder) loop() {
 	err := b.loadChallenge()
 	if err != nil {
 		log.Info("challenge not loaded: %s", err)
 	}
+	select {
+	case <-b.stop:
+		return
+	case <-b.initDone:
+	}
+	<-b.layerClock.AwaitLayer(1) // this doesn't block unless layer 1 hasn't arrived yet
 	for {
 		select {
 		case <-b.stop:
 			return
-		case layer := <-b.timer:
-			if !b.isSynced() {
-				b.log.Info("cannot create atx : not synced")
-				break
-			}
-
-			if atomic.LoadInt32(&b.initStatus) != InitDone {
-				b.log.Info("post is not initialized yet, not building Nipst")
-				break
-			}
-			if b.working {
-				break
-			}
-			b.working = true
-			go func() {
-				epoch := layer.GetEpoch(b.layersPerEpoch)
-				err := b.PublishActivationTx(epoch)
-				if err != nil {
-					if err == tooSoonErr {
-						b.log.Warning("cannot create atx in epoch %v: %v", epoch, err)
-					} else {
-						b.log.Error("cannot create atx in epoch %v: %v", epoch, err)
-					}
-					events.Publish(events.AtxCreated{false, "", uint64(epoch)})
-				}
-
-				b.finished <- layer
-			}()
-		case <-b.finished:
-			b.working = false
+		default:
```

Comment:
why leave the default?

Reply:
If I remove the `default:` it will block on the `stop` channel. By having a `default:` I check if `stop` has been closed, `return` if it did, and otherwise continue execution.